back/app/controllers/web_api/v1/projects_controller.rb

```diff
@@ -52,6 +52,73 @@ def index
     )
   end
 
+  # For use with 'Open to participation' homepage widget.
+  # Returns all published or archived projects that are visible to user
+  # and in an active participatory phase (where user can do something).
+  # Ordered by the end date of the current phase, soonest first (nulls last).
+  def index_projects_with_active_participatory_phase
+    # Projects user can see, with active participatory (not information) phase & include the phases.end_at column
+    # We could use the ProjectPermissionSevice step to filter for active phases, but doing it here is more efficient.
+    subquery = policy_scope(Project)
+      .joins('INNER JOIN admin_publications AS admin_publications ON admin_publications.publication_id = projects.id')
+      .where(admin_publications: { publication_status: 'published' })
+      .joins('INNER JOIN phases AS phases ON phases.project_id = projects.id')
+      .where(
+        'phases.start_at <= ? AND (phases.end_at >= ? OR phases.end_at IS NULL) AND phases.participation_method != ?',
+        Time.zone.now.to_fs(:db), Time.zone.now.to_fs(:db), 'information'
+      )
+      .select('projects.*, phases.end_at AS phase_end_at')
+
+    # Perform the SELECT DISTINCT on the outer query
+    @projects = Project
+      .from(subquery, :projects)
+      .distinct
+      .order('phase_end_at ASC NULLS LAST')
+      .preload(phases: { permissions: [:groups] })
+
+    # Projects user can participate in, or where such participation could (probably) be made possible by user
+    # (e.g. user not signed in).
+    # Unfortunately, this breaks the query chain, so we have to start a new one after this.
+    # Also, we will need the action descriptors again later, so we will store them.
+
+    # # Step 1: Create pairs of project ids and action descriptors.
+    # # Since this is the last filtering step, we will keep going
+    # # until we reach the limit required for pagination.
+    pagination_limit = calculate_pagination_limit
+
+    project_descriptor_pairs = @projects.each_with_object({}) do |project, acc|
+      service = Permissions::ProjectPermissionsService.new(
+        project, current_user, user_requirements_service: user_requirements_service
+      )
+      action_descriptors = service.action_descriptors
+
+      if Permissions::ProjectPermissionsService.participation_possible?(action_descriptors)
+        acc[project.id] = action_descriptors
+        break if acc.size >= pagination_limit
+      end
+    end
+
+    # # Step 2: Use these to filter out projects
+    project_ids = project_descriptor_pairs.keys
+    @projects = @projects.where(id: project_ids)
```

Comment:
Maybe it would be faster to do

```
@projects = Project.where(id: project_ids)
```

Because otherwise the whole query above from line 62-77 will be executed twice, whereas in this case that query will once be executed once + a lighter query where we just grab a few projects by id?

Reply:
Yup, indeed. 👍

Reply:
Except, when I check the n of queries, by editing the expected n to be some low n (e.g. 4) in [this test](https://github.com/CitizenLabDotCo/citizenlab/blob/307db8b96f0d24dc97548b82677d884d6fdd5ccd/back/spec/acceptance/projects_spec.rb#L1327), the n increases (from 29 to 32), when using your suggested approach.

So, it seems that maybe Rails is clever enough not to do the 2 separate queries. I'd have to check the diff between the 2 sets of queries that test lists, however, to be sure.

Reply:
Also, if we make your suggested change, we'd need to move the ordering query to after this query, as otherwise we 
will lose the ordering.

Reply:
@luucvanderzee Looking at the output from the test I mentioned, these 3 queries are added (in that test scenario) when using your suggested approach, at or near the end of the query chain...
```SQL
28) SELECT "permissions"."id" FROM "permissions" WHERE "permissions"."permission_scope_id" = $1 AND "permissions"."permission_scope_type" = $2; ["2f9ac3e2-09bf-4ff1-bbe1-30c10c993f13", "Phase"]  (0.833 ms)
...
30) SELECT MAX("phases"."end_at") FROM "phases" WHERE "phases"."project_id" = $1; ["fb4ae2e7-05cb-42b8-8fee-272ad2d3ed86"]  (1.077 ms)
...
-32) SELECT "permissions"."id" FROM "permissions" WHERE "permissions"."permission_scope_id" = $1 AND "permissions"."permission_scope_type" = $2; ["4e7b8e3b-591c-4f0a-b50b-55895e4c2b0c", "Phase"]  (1.144 ms)
```

Reply:
Possibly because it is a new query it loses some of the pre-loading here

Reply:
Indeed, I suspect Luuc's approach is best. I'll implement + move ordering query to after it.